### PR TITLE
min const generics

### PR DESCRIPTION
**Pull request by jon-chuang**
Replacement for https://github.com/arkworks-rs/algebra/pull/182
Mostly done, what's left is to review/decide on API issues/integrate with rest of codebase (like curves)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (main)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: ff/src/fields/arithmetic.rs [hunk 1]
```diff
@@ -6,240 +6,191 @@
...
+                BigInt::<N>(r)
             }
-            tmp.0 = r;
-            tmp
         }
     };
 }
-
-macro_rules! impl_field_square_in_place {
+// macro_rules! impl_deserialize_flags {
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:62]:
```suggestion

```

File: ff/src/fields/arithmetic.rs [hunk 2]
```diff
@@ -6,240 +6,191 @@
...
-                if i < $limbs - 1 {
-                    for j in 0..$limbs {
-                        if j > i {
-                            r[i + j] =
-                                mac_with_carry!(r[i + j], (self.0).0[i], (self.0).0[j], &mut carry);
-                        }
-                    }
-                    r[$limbs + i] = carry;
-                    carry = 0;
+// macro_rules! impl_deserialize_flags {
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:93]:
```suggestion

```

**Comment by Pratyush** [comment]
The macro infrastructure is kinda difficult to follow; is there a way to simplify it? Also, I think merging the `additive_ops` and `multiplicative_ops` macro complicates the macro impl  for only a small benefit. Could we revert that change?

File: ff/src/fields/models/fp.rs [hunk 3]
```diff
@@ -0,0 +1,833 @@
...
+        r
+    }
+}
+
+impl<P: FpParams<N>, const N: usize> FftField for Fp<P, N> {
+    type FftParams = P;
+
+    #[inline]
+    fn two_adic_root_of_unity() -> Self {
+        Fp::<P, N>(P::TWO_ADIC_ROOT_OF_UNITY, PhantomData)
```
Comment (Pratyush) [inline ff/src/fields/models/fp.rs:467]:
```suggestion
        Self::new(P::TWO_ADIC_ROOT_OF_UNITY)
```

File: ff/src/fields/models/fp.rs [hunk 4]
```diff
@@ -0,0 +1,833 @@
...
+    }
+
+    #[inline]
+    fn large_subgroup_root_of_unity() -> Option<Self> {
+        Some(Fp::<P, N>(P::LARGE_SUBGROUP_ROOT_OF_UNITY?, PhantomData))
```
Comment (Pratyush) [inline ff/src/fields/models/fp.rs:472]:
```suggestion
        Some(Self::new(P::LARGE_SUBGROUP_ROOT_OF_UNITY?))
```

File: ff/src/fields/models/fp.rs [hunk 5]
```diff
@@ -0,0 +1,833 @@
...
+    }
+
+    #[inline]
+    fn multiplicative_generator() -> Self {
+        Fp::<P, N>(P::GENERATOR, PhantomData)
```
Comment (Pratyush) [inline ff/src/fields/models/fp.rs:477]:
```suggestion
        Self::new(P::GENERATOR)
```

File: ff/src/fields/models/fp.rs [hunk 6]
```diff
@@ -0,0 +1,833 @@
...
+
+impl<P: FpParams<N>, const N: usize> Neg for Fp<P, N> {
+    type Output = Self;
+    #[inline]
+    #[must_use]
+    fn neg(self) -> Self {
+        if !self.is_zero() {
+            let mut tmp = P::MODULUS.clone();
+            tmp.sub_noborrow(&self.0);
+            Fp::<P, N>(tmp, PhantomData)
```
Comment (Pratyush) [inline ff/src/fields/models/fp.rs:703]:
```suggestion
            Self::new(tmp)
```

**Comment by jon-chuang** [comment]
> The macro infrastructure is kinda difficult to follow; is there a way to simplify it? Also, I think merging the `additive_ops` and `multiplicative_ops` macro complicates the macro impl for only a small benefit. Could we revert that change?

I've added some comments. 

I guess that given our usecase, I can simplify by making more defaults

**Comment by jon-chuang** [comment]
To be quite honest, the limitations of const means that const generics at present buys little imo, under the hood, one is more or less instantiating concretely over limb types within ff anyway. So it's unclear what one should have expected or might hope is yet to be improved.

**Comment by jon-chuang** [comment]
I would say that I did prefer the previous method of just implementing concretely via macros. Using const generics while requiring literals for unrolls means some unpleasant workarounds that make it harder to modify the code.

**Comment by huitseeker** [comment]
I would consider adding a check for the minimal rustc version supporting `min_const_generics.

In fact, I have this in a commit you can grab here:
https://github.com/huitseeker/algebra/commit/c69b1ed6a0505911e0335d63fc46f2eab204a057

Here's a sample output on an older compiler:
https://gist.github.com/cd63c76a4e53c2f4bbf19f9ca29f4eba

**Comment by Pratyush** [comment]
Superseded by #372 and #379

**Pull request closed without merging**